master/buildbot/status/web/forms.py

```diff
@@ -45,55 +93,67 @@ def decodeFromURL(self, value, encoding):
-                args[name] = [ self.decodeFromURL(arg, encoding) for arg in argl ]
+        args = self._decodeArgs(request, encoding)
 
         #Get builder info
         builder_status = None
         if args.has_key("builder_name") and len(args["builder_name"]) == 1:
             builder_status = status.getBuilder(self.decodeFromURL(args["builder_name"][0], encoding))
-            bm = self.getBuildmaster(request)
-
-            cxt['slaves'] = [s for s in builder_status.getSlaves() if s.isConnected()]
-            cxt['slaves'] = sorted(cxt['slaves'], key=attrgetter('friendly_name'))
-
-            #Get branches
-            encoding = getRequestCharset(request)
-            branches = [ b.decode(encoding) for b in args.get("branch", []) if b ]
-            cxt['branches'] = branches
-
-            return_page = ""
-            if args.has_key("return_page"):
-                return_page = args['return_page']
-                if not isinstance(return_page, basestring):
-                    return_page = args['return_page'][0]
-
-            #Add scheduler info
-            buildForceContext(cxt, request, self.getBuildmaster(request), builder_status.getName())
-
-            #URL to force page with return param
-            url = args['builder_url'][0]
-            url_parts = list(urlparse(url))
-
-            if len(url_parts) > 0 and len(return_page) > 0:
-                return_page = "&returnpage={0}".format(return_page)
-            else:
-                return_page = "?returnpage={0}".format(return_page)
-            cxt['return_page'] = return_page
-
-            url_parts[2] += "/force"
-            url_parts[4] += return_page
-            force_url = urlunparse(url_parts)
-            cxt['force_url'] = force_url
-
-            authz = self.getAuthz(request)
-            cxt['is_admin'] = yield authz.getUserAttr(request, 'is_admin', 0)
+            buildMaster = self.getBuildmaster(request)
+
+            cxt['slaves'] = self._getSlaves(builder_status)
+            cxt['branches'] = self._getBranches(args, request)
+
+            buildForceContext(cxt, request, buildMaster, builder_status.getName())
+
+            url_parts = list(urlparse(args['builder_url'][0]))
+            cxt['return_page'] = self._getReturnPage(args, url_parts)
+            cxt['force_url'] = self._getForceUrl(cxt['return_page'], url_parts)
+
+            cxt['is_admin'] = yield self._getIsAdmin(request)
+            cxt['rt_update'] = args
+            request.args = args
+
+            template = request.site.buildbot_service.templates.get_template("force_build_dialog.html")
+            defer.returnValue(template.render(**cxt))
+
+        else:
+            page = ErrorPage(INTERNAL_SERVER_ERROR, "Missing parameters", "Not all parameters were given")
+            defer.returnValue(page.render(request))
+
+class RebuildDialogPage(BuildDialogPage, HtmlResource):
+    pageTitle = "Rebuild"
+
+    def decodeFromURL(self, value, encoding):
```

Comment:
What do you think about moving this method to `BuildDialogPage` class ?

Reply:
👍 yeah, it is even duplicated in both classes